src/compiler/checker.ts

```diff
@@ -17651,12 +17652,51 @@ namespace ts {
             }
         }
 
+        function getOptionalTypeRecursive(type: Type) {
+            if ((type as ResolvedType).properties) {
+                const rType = type as ResolvedType;
+                const propCount = rType.properties.length;
+
+                if (rType.flags & TypeFlags.Object) {
+                  for (let i = 0; i < propCount; i++) {
+                    const prop = rType.properties[i] as SymbolLinks;
+                    let innerType;
+                    if (prop.bindingElement && prop.bindingElement.initializer) {
+                        innerType = getOptionalTypeRecursive(prop.type);
+                        innerType = getOptionalType(innerType);
+                    }
+                    else if (prop.type.symbol) {
+                        innerType = getTypeOfParameter(prop.type.symbol);
+                    }
+                    else {
+                        innerType = prop.type;
+                    }
+                    prop.type = innerType;
+                    rType.properties[i] = prop as Symbol;

```

Comment:
You need to make a new type instead of overwriting the properties of an existing type. Somebody else might expect the properties of that type to remain the same.

Reply:
Same with symbol; you can't just update the type of an existing symbol.

Reply:
That seems reasonable. I should wipe Haskell out of my head for good. :-D